compiler/vectorized_test.go

```diff
@@ -0,0 +1,150 @@
+		fn           string
+		vectorizable bool
+		allocated    int64
+		maxAllocated int64
+		inType       semantic.MonoType
+		input        map[string]interface{}
+		want         map[string]interface{}
+		skipComp     bool
+	}{
+		{
+			name:         "field access",
+			fn:           `(r) => ({c: r.a, d: r.b})`,
+			vectorizable: true,
+			inType: semantic.NewObjectType([]semantic.PropertyType{
+				{Key: []byte("r"), Value: semantic.NewObjectType([]semantic.PropertyType{
+					{Key: []byte("a"), Value: semantic.NewVectorType(semantic.BasicInt)},
+					{Key: []byte("b"), Value: semantic.NewVectorType(semantic.BasicInt)},
+				})},
+			}),
+			input: map[string]interface{}{
+				"r": map[string]interface{}{
+					"a": []interface{}{int64(1)},
+					"b": []interface{}{int64(2)},
+				},
+			},
+			want: map[string]interface{}{
+				"c": []interface{}{int64(1)},
+				"d": []interface{}{int64(2)},
+			},
+		},
+		{
+			name:         "extend record",
+			fn:           `(r) => ({r with b: r.a})`,
+			vectorizable: true,
+			inType: semantic.NewObjectType([]semantic.PropertyType{
+				{Key: []byte("r"), Value: semantic.NewObjectType([]semantic.PropertyType{
+					{Key: []byte("a"), Value: semantic.NewVectorType(semantic.BasicFloat)},
+				})},
+			}),
+			input: map[string]interface{}{
+				"r": map[string]interface{}{
+					"a": []interface{}{1.2},
+				},
+			},
+			want: map[string]interface{}{
+				"a": []interface{}{1.2},
+				"b": []interface{}{1.2},
+			},
+		},
+		{
+			name:         "no binary expressions",
+			fn:           `(r) => ({c: r.a + r.b})`,
+			vectorizable: false,
+			skipComp:     true,
+		},
+		{
+			name:         "no literals",
+			fn:           `(r) => ({r with c: "count"})`,
+			vectorizable: false,
+			skipComp:     true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			mem := &memory.Allocator{}
+			checked := arrow.NewCheckedAllocator(mem)
+			pkg, err := runtime.AnalyzeSource(tc.fn)
+			if err != nil {
+				t.Fatalf("unexpected error: %s", err)
+			}
+
+			stmt := pkg.Files[0].Body[0].(*semantic.ExpressionStatement)
+			fn := stmt.Expression.(*semantic.FunctionExpression)
+
+			if tc.vectorizable {
+				if fn.Vectorized == nil {
+					t.Fatal("Expected to find vectorized node, but found none")
+				}
+			} else {
+				if fn.Vectorized != nil {
+					t.Fatal("Vectorized node is populated when it should be nil")
+				}
+			}
+
+			if tc.skipComp {
+				return
+			}
+
+			f, err := compiler.Compile(nil, fn, tc.inType)
+			if err != nil {
+				t.Fatalf("unexpected error: %s", err)
+			}
+			input := vectorizedObjectFromMap(tc.input, mem)
+			got, err := f.Eval(context.TODO(), input)
+			if err != nil {
+				t.Fatalf("unexpected error: %s", err)
+			}
+
+			want := vectorizedObjectFromMap(tc.want, &memory.Allocator{})
+			if !cmp.Equal(want, got, CmpOptions...) {
+				t.Errorf("unexpected value -want/+got\n%s", cmp.Diff(want, got, CmpOptions...))
+			}
+
+			checked.AssertSize(t, 0)
```

Comment:
This test would definitely benefit from the `debug` build tag that would catch too many calls to `Release`.